### PR TITLE
Multi hop

### PR DESCRIPTION
**Pull request by Priscilla-AL**
WHAT DOES THIS PR DO?
Creates an endpoint for computing calibrated values using the multi-hop approach

WHAT JIRA STORY/TASK IS RELATED TO THIS PR?
https://airqoteam.atlassian.net/jira/software/projects/PLAT/boards/41?assignee=60095b78dfb0c70069357e54&selectedIssue=PLAT-523 

WHAT IS THE LINK TO THE PR BRANCH
https://github.com/airqo-platform/AirQo-api/tree/Multi-hop

HOW DO I TEST OUT THIS PR?
cd AirQo-api/src/calibrate-multi-hop
python -m venv [DESTINATION_FOLDER]
source [DESTINATION_FOLDER]/Scripts/activate
pip install -r requirements.txt
flask run

WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:
[POST] http://localhost:4001/api/v1/calibrate/multihop

Sample Input body -
{
"datetime":"2020-09-06 12:56:59+00:00",
"raw_values": [
{
"raw_value": 100,
"sensor_id": 832251
},
{
"raw_value": 22,
"sensor_id": 82720
},
{
"raw_value": 29,
"sensor_id": 870144
}
]
}

IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?
N/A

ARE THERE ANY RELATED PRs?
N/A

### CODE REVIEW

**Comment by bobturneruk** [comment]
Should be rebased on https://github.com/airqo-platform/AirQo-api/pull/343 after merge.

**Comment by bobturneruk** [comment]
https://github.com/airqo-platform/AirQo-api/pull/343 is now merged.

**Comment by bobturneruk** [comment]
Hi @Priscilla-AL - I've refined this a bit, reducing the requirements and drawing Mike's code from a different place. I've go the the point where I get this when I make the API call that I've specified in the new [readme.md](https://github.com/airqo-platform/AirQo-api/blob/Multi-hop/src/calibrate-multi-hop/readme.md):

```
TypeError: compute_simple_predictions() missing 2 required positional arguments:
				'allcals' and 'delta'
```

Can you let me know if this is what you might expect, please? Do you have another way of testing the code? Maybe I've misinterpreted what I'm supposed to be doing!

**Review by Priscilla-AL — COMMENTED** [review]
(env) C:\Users\User\Documents\Projects\AirQo-api\src\calibrate-multi-hop>pip install -r requirements.txt
Collecting pandas>=1.2.4 (from -r requirements.txt (line 1))
  Could not find a version that satisfies the requirement pandas>=1.2.4 (from -r requirements.txt (line 1)) (from versions: 0.1, 0.2b0, 0.2b1, 0.2, 0.3.0b0, 0.3.0b2, 0.3.0, 0.4.0, 0.4.1, 0.4.2, 0.4.3, 0.5.0, 0.6.0, 0.6.1, 0.7.0rc1, 0.7.0, 0.7.1, 0.7.2, 0.7.3, 0.8.0rc1, 0.8.0rc2, 0.8.0, 0.8.1, 0.9.0, 0.9.1, 0.10.0, 0.10.1, 0.11.0, 0.12.0, 0.13.0, 0.13.1, 0.14.0, 0.14.1, 0.15.0, 0.15.1, 0.15.2, 0.16.0, 0.16.1, 0.16.2, 0.17.0, 0.17.1, 0.18.0, 0.18.1, 0.19.0rc1, 0.19.0, 0.19.1, 0.19.2, 0.20.0rc1, 0.20.0, 0.20.1, 0.20.2, 0.20.3, 0.21.0rc1, 0.21.0, 0.21.1, 0.22.0, 0.23.0rc2, 0.23.0, 0.23.1, 0.23.2, 0.23.3, 0.23.4, 0.24.0rc1, 0.24.0, 0.24.1, 0.24.2, 0.25.0rc0, 0.25.0, 0.25.1, 0.25.2, 0.25.3, 1.0.0rc0, 1.0.0, 1.0.1, 1.0.2, 1.0.3, 1.0.4, 1.0.5, 1.1.0rc0, 1.1.0, 1.1.1, 1.1.2, 1.1.3, 1.1.4, 1.1.5)
No matching distribution found for pandas>=1.2.4 (from -r requirements.txt (line 1))
You are using pip version 10.0.1, however version 21.1.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.

How about removing the versions

**Comment by Priscilla-AL** [comment]
> Hi @Priscilla-AL - I've refined this a bit, reducing the requirements and drawing Mike's code from a different place. I've go the the point where I get this when I make the API call that I've specified in the new [readme.md](https://github.com/airqo-platform/AirQo-api/blob/Multi-hop/src/calibrate-multi-hop/readme.md):
> 
> ```
> TypeError: compute_simple_predictions() missing 2 required positional arguments:
> 				'allcals' and 'delta'
> ```
> 
> Can you let me know if this is what you might expect, please? Do you have another way of testing the code? Maybe I've misinterpreted what I'm supposed to be doing!


allcals and delta are defined in models/calibrate.py.  It should run fine.
We can test it by checking docker image size  I guess.

**Comment by bobturneruk** [comment]
Thanks @Priscilla-AL  - 1.2.4 of pandas is on pypi https://pypi.org/project/pandas/ What python version are you using, please? I'm using 3.7 and I've mostly been testing using docker.

**Comment by Priscilla-AL** [comment]
> Thanks @Priscilla-AL - 1.2.4 of pandas is on pypi https://pypi.org/project/pandas/ What python version are you using, please? I'm using 3.7 and I've mostly been testing using docker.

Am also using 3.7. 0, but testing as described above

**Comment by bobturneruk** [comment]
Do you get the error when building the docker image, then?

**Comment by Priscilla-AL** [comment]
> Do you get the error when building the docker image, then?

It works after removing versions from pandas.

**Comment by bobturneruk** [comment]
This is odd - I'd've thought that there wouldn't be any difference between building docker images on two different machines. Can you let me know what version of pandas works for you, please, @Priscilla-AL? I think it's probably wise to specify a minimum. In the meantime, I now seem to be getting a different error on the API call, so I'll look at that.

**Comment by bobturneruk** [comment]
Current error in response to API call specified in readme.md is:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2069, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2054, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib/python3.7/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2051, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1501, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.7/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1499, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1485, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/app/controllers/calibrate.py", line 50, in calibrate
    value = calibrateModel.calibrate_sensor_raw_data(datetime, raw_value.get('sensor_id'), raw_value.get('raw_value'))
  File "/app/models/calibrate.py", line 56, in calibrate_sensor_raw_data
    res,scale,preds,key = sp.compute_simple_predictions(testX,self.allcals,delta)
TypeError: compute_simple_predictions() missing 2 required positional arguments: 'allcals' and 'delta'
```

**Comment by bobturneruk** [comment]
@Priscilla-AL - I've got rid of the pandas version requirement, but I still get the same error on an API call using the instructions in readme.md
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2069, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2054, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib/python3.7/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2051, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1501, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.7/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1499, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1485, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/app/controllers/calibrate.py", line 50, in calibrate
    value = calibrateModel.calibrate_sensor_raw_data(datetime, raw_value.get('sensor_id'), raw_value.get('raw_value'))
  File "/app/models/calibrate.py", line 56, in calibrate_sensor_raw_data
    res,scale,preds,key = sp.compute_simple_predictions(testX,self.allcals,delta)
TypeError: compute_simple_predictions() missing 2 required positional arguments: 'allcals' and 'delta'
```
Do you get the same error? Is there anything you can think of that I could change? I'll need to get past this to do more analysis / optimisation.

**Comment by Priscilla-AL** [comment]
> Do you get the error when building the docker image, then?

No, the docker image was built successfully

**Comment by bobturneruk** [comment]
Sorry - I mean do you get the above error when making an API call as detailed here? https://github.com/airqo-platform/AirQo-api/blob/Multi-hop/src/calibrate-multi-hop/readme.md That's when I get it. Great that the docker image is building!

**Comment by Priscilla-AL** [comment]
> Sorry - I mean do you get the above error when making an API call as detailed here? https://github.com/airqo-platform/AirQo-api/blob/Multi-hop/src/calibrate-multi-hop/readme.md That's when I get it. Great that the docker image is building!

Yes, although I think it is coming from the simple.py. I may need to modify a function. Will take a look at it

**Review by Priscilla-AL — COMMENTED** [review]
@bobturneruk I have failed to resolve issues with the end point. How about a short working meeting. 30 minutes

**Comment by bobturneruk** [comment]
Hi! Yes. That sounds good. It's a public holiday here today. I'll be in
touch tomorrow.

**Comment by bobturneruk** [comment]
I'm now getting:
```
[
    {
        "Sensor_id": null,
        "calibrated_value": null
    },
    {
        "Sensor_id": null,
        "calibrated_value": null
    },
    {
        "Sensor_id": null,
        "calibrated_value": null
    }
]
```
when I follow through the readme. Most odd!

**Comment by bobturneruk** [comment]
@Priscilla-AL - I've rebased this branch on `staging`, rebuilt container and re-tested with postman. Same response as above. Do you get the same behaviour, please?

**Comment by bobturneruk** [comment]
I get the same if I don't use Docker.

File: src/calibrate-multi-hop/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+pandas
+numpy
+pymongo>=3.11.3
+Flask_Cors>=3.0.10
+Flask>=1.1.2
+Flask_PyMongo>=2.3.0
+python-dotenv>=0.17.1
+git+https://github.com/airqo-platform/AirQo-modules.git@calibration-simple#egg=simple_calibration&subdirectory=calibration/simple_calibration
```
Comment (Baalmart) [inline src/calibrate-multi-hop/requirements.txt:8]: Hi @bobturneruk and @Priscilla-AL 
Lovely, so looks like this is where the package is consumed from.
Could we explore publishing these packages? How is the process like? Smooth or complicated for our usecase?
A little search made me land on this: https://www.codementor.io/@ajayagrawal295/how-to-publish-your-own-python-package-12tbhi20tf

**Review by Baalmart — COMMENTED** [review]
Looks good. Shall leave more detailed comments in the airqo-modules repo [PR-11](https://github.com/airqo-platform/AirQo-modules/pull/11).....

File: src/calibrate-multi-hop/requirements.txt (hunk 1 above)
Comment (bobturneruk) [reply src/calibrate-multi-hop/requirements.txt:8]: It's fine - I'm very familiar with it. It should work well.

**Comment by bobturneruk** [comment]
Note to self: Currently getting:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2088, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib/python3.7/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2070, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1515, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.7/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1513, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1499, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/app/controllers/calibrate.py", line 47, in calibrate
    calibrateModel = cb.Calibrate()
  File "/app/models/calibrate.py", line 19, in __init__
    encounters = pickle.load(pickle_file)
  File "pandas/_libs/internals.pyx", line 572, in pandas._libs.internals.BlockManager.__cinit__
    
TypeError: __cinit__() takes at least 2 positional arguments (0 given)
```
When testing as per readme. I think this may be a `pandas` version 1.3.0 issue.

**Comment by bobturneruk** [comment]
perhaps changing `pickle.load` to `pd.read_pickle` will fix

**Comment by bobturneruk** [comment]
This currently passes tests specified in the microservice readme when run using docker. https://github.com/airqo-platform/AirQo-modules/pull/11 should be merged and https://github.com/airqo-platform/AirQo-api/blob/084e3039f34dfe7406f8b30ff59a479c7b1988bb/src/calibrate-multi-hop/requirements.txt#L8 updated before this is merged.

File: src/calibrate-multi-hop/app.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+# import blue prints
+
+_logger = logging.getLogger(__name__)
+
+app = Flask(__name__)
+
+# Allow cross-brower resource sharing
+CORS(app)
+
+app.config["MONGO_URI"] = os.getenv("MONGO_URI")
```
Comment (Baalmart) [inline src/calibrate-multi-hop/app.py:19]: please share the .env for this micro in DM perhaps? Thanks. @bobturneruk

File: src/calibrate-multi-hop/controllers/calibrate.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+from models import calibrate as cb
+
+
+calibrate_bp = Blueprint('calibrate_bp', __name__)
+
+
+MONGO_URI = os.getenv('MONGO_URI')
+client = MongoClient(MONGO_URI)
+db = client['airqo_netmanager_staging_airqo']
+col = db['calibration_ratios']
```
Comment (Baalmart) [inline src/calibrate-multi-hop/controllers/calibrate.py:18]: I sense a mismatch here, the default ENV in the docker file is production but the DB here is hard corded to staging......could we review this? @bobturneruk @Priscilla-AL

File: src/calibrate-multi-hop/controllers/calibrate.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+@calibrate_bp.route(api.route['ratios'], methods=['POST', 'GET'])
```
Comment (Baalmart) [inline src/calibrate-multi-hop/controllers/calibrate.py:20]: I am one of those who is cautious about the usage of these HTTP verbs, could there be any reason why we allow both POST and GET on the same endpoint? Just need some enlightenment....could it be that we want to trigger 2 different actions based on HTTP verbs used in the request? Or do we just want to use both POST and GET on the same implementation? Just curious....

Comment (Baalmart) [reply src/calibrate-multi-hop/controllers/calibrate.py:20]: What could be the drawback of using/allowing one? That is, either POST only or GET only?

File: src/calibrate-multi-hop/models/calibrate.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        G,allsp,allcals,allcallists,allpopts,allpcovs,allpoptslists = sp.compute_simple_calibration(X,Y,delta,refsensor)
+        return allcals, myDict_id
+   
+  
+    def calibrate_sensor_raw_data(self, datetime, sensor_id, raw_value):
+        
+        delta = 24*7
+
+        datetime = pd.Timestamp(datetime, tz='UTC')
+        time = (datetime-pd.Timestamp('2020-07-15',tz='UTC')).total_seconds()/3600
```
Comment (Baalmart) [inline src/calibrate-multi-hop/models/calibrate.py:47]: Hi @bobturneruk 
Could we have a short chat (Hangouts call) early next week to review this implementation? I might appreciate some aspects on a call perhaps.....potential candidates on that suggested **short** working could be @NoahNsimbe and @ja-odur .....thanks.....

File: src/calibrate-multi-hop/.gitignore [hunk 6]
```diff
@@ -0,0 +1,10 @@
+.env
+__pycache__/
+env/
+__init__test.py
+.gitkeep
+.vscode
+.DS_Store
+airqo-250220-5149c2aac8f2.json
```
Comment (Baalmart) [inline src/calibrate-multi-hop/.gitignore:8]: Hi @bobturneruk 
I don't seem to have this json file, the one shared in our channel by @Priscilla-AL has this name: **AirQo-e37846081569.json**

**Review by Baalmart — COMMENTED** [review]
Thanks @bobturneruk , interesting additions and neat work!
I am unable to run this new microservice right now......just left a couple of comments/suggestions for your review.

Thanks!

File: src/calibrate-multi-hop/app.py (hunk 2 above)
Comment (bobturneruk) [reply src/calibrate-multi-hop/app.py:19]: The environment variables should be fully specified by the command line arguments in the readme (https://github.com/airqo-platform/AirQo-api/blob/Multi-hop/src/calibrate-multi-hop/readme.md#to-build-and-run-with-docker-desktop). Are there any we've missed, @Baalmart?

File: src/calibrate-multi-hop/controllers/calibrate.py (hunk 3 above)
Comment (bobturneruk) [reply src/calibrate-multi-hop/controllers/calibrate.py:18]: Good spot - I guess this should be changed @Priscilla-AL?

File: src/calibrate-multi-hop/controllers/calibrate.py (hunk 4 above)
Comment (bobturneruk) [reply src/calibrate-multi-hop/controllers/calibrate.py:20]: Not sure - @Priscilla-AL?

File: src/calibrate-multi-hop/.gitignore (hunk 6 above)
Comment (bobturneruk) [reply src/calibrate-multi-hop/.gitignore:8]: I don't have a local json file at all for this.

Comment (bobturneruk) [reply src/calibrate-multi-hop/.gitignore:8]: Is one needed for this microservice, for credentials?

**Comment by bobturneruk** [comment]
@ja-odur and @danielogen - please have a look at this and help us to make changes so it will work better as a microservice on the platform!

**Comment by danielogen** [comment]
> @ja-odur and @danielogen - please have a look at this and help us to make changes so it will work better as a microservice on the platform!

Hi @bobturneruk and @Priscilla-AL,

Kindly check the new changes added to this microservice. I have configured and tested docker as well. You can try this on staging by running: `docker build -t test-multihop -f Docker.stage .`, `docker run -it -p 4001:4001 test-multihop`.
Make sure you have the `.env` provided on slack.
cc: @Baalmart 

Just out of curiosity,  I'm not sure why the date is hard coded (refer to the screenshot below)
Screen Shot 2021-07-25 at 9 19 27 PM

**Comment by bobturneruk** [comment]
I've made a couple of suggestions. I'm currently getting:

```
{
    "message": "please specify the organization name. Refer to the API documentation for details.",
    "success": false
}
```

when I follow the readme instructions. Do I need to modify the `json` payload in the readme? cc @Priscilla-AL @danielogen

File: src/calibrate-multi-hop/requirements.txt (hunk 1 above)
Comment (Baalmart) [reply src/calibrate-multi-hop/requirements.txt]: Awesome, that will be great! Looking forward to trying it out. Perhaps a short workshop of sorts? @bobturneruk

**Pull request closed without merging**